main-command/src/main/scala/sbt/internal/client/BspClient.scala

```diff
@@ -18,24 +18,39 @@ import scala.sys.process.Process
 import scala.util.control.NonFatal
 
 class BspClient private (sbtServer: Socket) {
+  private var terminated = false
 
   private def transferTo(input: InputStream, output: OutputStream): Unit = {
     val buffer = Array.ofDim[Byte](1024)
     while (true) {
       val size = input.read(buffer)
-      output.write(buffer, 0, size)
-      output.flush()
+      if (size == -1) synchronized {
+        output.close()
+        terminated = true
+        notify()
+      } else {
+        output.write(buffer, 0, size)
```

Comment:
All of the io calls (`input.read`, `output.write` and `output.flush`) could throw an `IOException`. For example, if you `kill -9` the server, I think `input.read` will throw an exception rather than return `-1`. My guess is that you want to exit if that happens so I'd recommend adding a `try` / `catch` block that sets terminated to true and notifies the main thread if an unexpected error occurs.

Reply:
(this by the way is why I think scala's ignoring checked exceptions from java was a horrible design decision).

Reply:
From what I have experienced the socket would return `-1` even if the server process is abruptly killed. But that may depends on the implementation which itself depends on the OS and JVM. So I think you are right, it is a lot safer to catch and terminate the process in case of exception.